Source/BugsnagSessionTracker.m

```diff
@@ -87,34 +92,39 @@ - (void)incrementHandledError {
 }
 
 - (void)send {
-    @synchronized (self.sessionStore) {
-        NSMutableArray *sessions = [NSMutableArray new];
-        NSArray *fileIds = [self.sessionStore fileIds];
+    NSArray *fileIds = [self.sessionStore fileIds];
 
-        for (NSDictionary *dict in [self.sessionStore allFiles]) {
-            [sessions addObject:[[BugsnagSession alloc] initWithDictionary:dict]];
-        }
-        BugsnagSessionTrackingPayload *payload = [[BugsnagSessionTrackingPayload alloc] initWithSessions:sessions];
-
-        if (payload.sessions.count > 0) {
-            [self.apiClient sendData:payload
-                         withPayload:[payload toJson]
-                               toURL:self.config.sessionURL
-                             headers:self.config.sessionApiHeaders
-                        onCompletion:^(id data, BOOL success, NSError *error) {
-
-                            if (success && error == nil) {
-                                NSLog(@"Sent sessions to Bugsnag");
-
-                                for (NSString *fileId in fileIds) {
-                                    [self.sessionStore deleteFileWithId:fileId];
-                                }
-                            } else {
-                                NSLog(@"Failed to send sessions to Bugsnag: %@", error);
+    if (fileIds.count <= 0) {
+        return;
+    }
+
+    dispatch_semaphore_t requestSemaphore = dispatch_semaphore_create(0);
+    NSMutableArray *sessions = [NSMutableArray new];
+
+    for (NSDictionary *dict in [self.sessionStore allFiles]) {
+        [sessions addObject:[[BugsnagSession alloc] initWithDictionary:dict]];
+    }
+    BugsnagSessionTrackingPayload *payload = [[BugsnagSessionTrackingPayload alloc] initWithSessions:sessions];
+
+    if (payload.sessions.count > 0) {
+        [self.apiClient sendData:payload
+                     withPayload:[payload toJson]
+                           toURL:self.config.sessionURL
+                         headers:self.config.sessionApiHeaders
+                    onCompletion:^(id data, BOOL success, NSError *error) {
+                        if (success && error == nil) {
+                            NSLog(@"Sent sessions to Bugsnag");
+
+                            for (NSString *fileId in fileIds) {
+                                [self.sessionStore deleteFileWithId:fileId];
                             }
-                        }];
-        }
+                        } else {
+                            NSLog(@"Failed to send sessions to Bugsnag: %@", error);
+                        }
+                        dispatch_semaphore_signal(requestSemaphore);
+                    }];
     }
+    dispatch_semaphore_wait(requestSemaphore, DISPATCH_TIME_FOREVER);
```

Comment:
What happens here if `payload.sessions.count == 0` (and the previous block is never triggered)? Might need to be accounted for to prevent a deadlock.

Reply:
Addressed by sending a signal to the semaphore immediately in this case.